### PR TITLE
CSP : Check policy in CreateInputResource. Avoid modifying inline things and combining when appropriate.

### PR DESCRIPTION
**Pull request by morlovich**

### CODE REVIEW

**Review by oschaaf — APPROVED** [review]
This LGTM! 

One thought: do we need to decline rewriting domains in the domain rewrite filter if we foresee we could violate CSP?

Also, probably related to an earlier PR, but I do wonder about an error on master in the build which looks related: http://ci.onpagespeed.com/builds/mod_pagespeed/8a19c5d0471db3d5c86960dcec8e75455a316874/ubuntu-1404-x64-checkin-test-tmpdir/instaweb.26045/pagespeed_automatic_smoke_test.log

**Comment by morlovich** [comment]
Thanks for point that out. That failure looks like I forgot to add an .a for an intermediate target to PSOL.a makefile; I think I'll just roll that in here.

File: net/instaweb/rewriter/css_filter_test.cc [hunk 1]
```diff
@@ -2427,6 +2427,46 @@ TEST_F(CssFilterTest, AbsolutifyServingFallback) {
...
+  const char kCsp[] = "<meta http-equiv=\"Content-Security-Policy\" "
+                      "content=\"style-src */styles/ \">";
+  const char kCss[] = "<style>* { display: stylish; }</style>";
+  const char kStyledDiv[] = "<div style='background-color: #f00; '/>";
+
+  ValidateExpected(
+      "inline_css",
+      StrCat(kCsp, kCss, kStyledDiv),
+      StrCat(kCsp, kCss,
+             "<!--Avoiding modifying inline style with CSP present-->",
```
Comment (jmarantz) [inline net/instaweb/rewriter/css_filter_test.cc:2465]: WDYT about moving the string constant in the impl file into a class static and referencing it here?

File: net/instaweb/rewriter/js_combine_filter_test.cc [hunk 2]
```diff
@@ -1300,4 +1300,44 @@ TEST_F(JsCombineFilterTest, MapRewriteDomainAccrossDirs) {
   EXPECT_GT(lru_cache()->num_inserts(), 0);  // Actually did work.
 }
 
+TEST_F(JsCombineFilterTest, Csp) {
+  // Can't use EnableDebug since fixture uses SoftEnableFilterForTesting.
+  options()->ClearSignatureForTesting();
+  options()->SoftEnableFilterForTesting(RewriteOptions::kDebug);
+  server_context()->ComputeSignature(options());
+
+  const char kCsp[] = "<meta http-equiv=\"Content-Security-Policy\" "
```
Comment (jmarantz) [inline net/instaweb/rewriter/js_combine_filter_test.cc]: static const char kCsp[] = ...

File: net/instaweb/rewriter/js_combine_filter_test.cc [hunk 3]
```diff
@@ -1300,4 +1300,44 @@ TEST_F(JsCombineFilterTest, MapRewriteDomainAccrossDirs) {
...
+}
+
+TEST_F(JsCombineFilterTest, Csp2) {
+  // Can't use EnableDebug since fixture uses SoftEnableFilterForTesting.
+  options()->ClearSignatureForTesting();
+  options()->SoftEnableFilterForTesting(RewriteOptions::kDebug);
+  server_context()->ComputeSignature(options());
+
+  // This one has unsafe-eval, so we can work.
+  const char kCsp[] = "<meta http-equiv=\"Content-Security-Policy\" "
```
Comment (jmarantz) [inline net/instaweb/rewriter/js_combine_filter_test.cc]: static

File: net/instaweb/rewriter/css_filter_test.cc [hunk 4]
```diff
@@ -2427,6 +2427,46 @@ TEST_F(CssFilterTest, AbsolutifyServingFallback) {
...
 }
 
+TEST_F(CssFilterTest, BasicCsp) {
+  EnableDebug();
+  SetResponseWithDefaultHeaders("styles/a.css",
+                                kContentTypeCss, kInputStyle, 100);
+  SetResponseWithDefaultHeaders("uploads/sneaky.png",
+                                kContentTypeCss, kInputStyle, 100);
+
+  const char kCsp[] = "<meta http-equiv=\"Content-Security-Policy\" "
```
Comment (jmarantz) [inline net/instaweb/rewriter/css_filter_test.cc]: static const char kCsp[] = ...

File: net/instaweb/rewriter/css_filter_test.cc [hunk 5]
```diff
@@ -2427,6 +2427,46 @@ TEST_F(CssFilterTest, AbsolutifyServingFallback) {
...
+             "<!--The preceding resource was not rewritten "
+             "because CSP disallows its fetch-->"));
+}
+
+TEST_F(CssFilterTest, InlineCsp) {
+  EnableDebug();
+  options()->ClearSignatureForTesting();
+  options()->EnableFilter(RewriteOptions::kRewriteStyleAttributes);
+
+  const char kCsp[] = "<meta http-equiv=\"Content-Security-Policy\" "
```
Comment (jmarantz) [inline net/instaweb/rewriter/css_filter_test.cc]: static const char kCsp here & below

File: net/instaweb/rewriter/javascript_filter_test.cc [hunk 6]
```diff
@@ -1410,6 +1414,49 @@ TEST_P(JavascriptFilterTest, ContentTypeValidation) {
...
+TEST_P(JavascriptFilterTest, BasicCsp) {
+  InitFilters();
+  EnableDebug();
+
+  SetResponseWithDefaultHeaders(
+      "scripts/a.js", kContentTypeJavascript, kJsData, 100);
+  SetResponseWithDefaultHeaders(
+      "uploads/sneaky.png", kContentTypeJavascript, kJsData, 100);
+
+  const char kCsp[] =
```
Comment (jmarantz) [inline net/instaweb/rewriter/javascript_filter_test.cc]: static

**Review by jmarantz — APPROVED** [review]
awesome stuff...just some trivial string comments.

File: net/instaweb/rewriter/css_filter_test.cc (hunk 4 above)
Comment (morlovich) [reply net/instaweb/rewriter/css_filter_test.cc]: Done

File: net/instaweb/rewriter/css_filter_test.cc (hunk 5 above)
Comment (morlovich) [reply net/instaweb/rewriter/css_filter_test.cc]: Done

File: net/instaweb/rewriter/js_combine_filter_test.cc (hunk 2 above)
Comment (morlovich) [reply net/instaweb/rewriter/js_combine_filter_test.cc]: Done

File: net/instaweb/rewriter/js_combine_filter_test.cc (hunk 3 above)
Comment (morlovich) [reply net/instaweb/rewriter/js_combine_filter_test.cc]: Done

File: net/instaweb/rewriter/javascript_filter_test.cc (hunk 6 above)
Comment (morlovich) [reply net/instaweb/rewriter/javascript_filter_test.cc]: Done

File: net/instaweb/rewriter/css_filter_test.cc (hunk 1 above)
Comment (morlovich) [reply net/instaweb/rewriter/css_filter_test.cc:2465]: typo'ing wouldn't cause it to fail to diagnose a bug, and I think it's more readable like this...

**Comment by morlovich** [comment]
csp-integrate3 updated to proper base, but it looks to be in a desperate need of tests, so not ready for review yet.

**Pull request merged**